// friendly-pow/node/puzzle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPuzzleTTL = exports.combineSolutions = exports.getPuzzleSolverInputs = exports.PUZZLE_USER_DATA_MAX_LENGTH = exports.PUZZLE_USER_DATA_OFFSET = exports.PUZZLE_NONCE_OFFSET = exports.PUZZLE_DIFFICULTY_OFFSET = exports.NUMBER_OF_PUZZLES_OFFSET = exports.PUZZLE_EXPIRY_OFFSET = exports.PUZZLE_VERSION_OFFSET = exports.APP_ID_OFFSET = exports.ACCOUNT_ID_OFFSET = exports.PUZZLE_TIMESTAMP_OFFSET = void 0;
const encoding_1 = require("./encoding");
const constants_1 = require("./constants");
exports.PUZZLE_TIMESTAMP_OFFSET = 0;
exports.ACCOUNT_ID_OFFSET = 4;
exports.APP_ID_OFFSET = 8;
exports.PUZZLE_VERSION_OFFSET = 12;
exports.PUZZLE_EXPIRY_OFFSET = 13;
exports.NUMBER_OF_PUZZLES_OFFSET = 14;
exports.PUZZLE_DIFFICULTY_OFFSET = 15;
exports.PUZZLE_NONCE_OFFSET = 24;
exports.PUZZLE_USER_DATA_OFFSET = 32;
exports.PUZZLE_USER_DATA_MAX_LENGTH = 32;
function getPuzzleSolverInputs(puzzleBuffer, numPuzzles) {
    const startingPoints = [];
    for (let i = 0; i < numPuzzles; i++) {
        const input = new Uint8Array(constants_1.CHALLENGE_SIZE_BYTES);
        input.set(puzzleBuffer);
        input[120] = i;
        startingPoints.push(input);
    }
    return startingPoints;
}
exports.getPuzzleSolverInputs = getPuzzleSolverInputs;
/**
 * Combine multiple solutions (8 byte values) into a single array
 * @param solutions
 */
function combineSolutions(solutions) {
    const combined = new Uint8Array(solutions.length * 8);
    for (let i = 0; i < solutions.length; i++) {
        combined.set(solutions[i], i * 8);
    }
    return combined;
}
exports.combineSolutions = combineSolutions;
/**
 * Time in seconds the puzzle is valid for.
 * @param puzzleBuffer
 */
function getPuzzleTTL(puzzleBuffer) {
    return encoding_1.expiryToDurationInSeconds(puzzleBuffer[exports.PUZZLE_EXPIRY_OFFSET]);
}
exports.getPuzzleTTL = getPuzzleTTL;
